### PR TITLE
Add a cheat to allow dual-wielding all weapons

### PR DESCRIPTION
**Pull request by throwaway29292**
The reason, why i have added this to the cheats is because it works both in SP as well as MP.

I couldn't set a custom name so i have chosen the name "(Two-Handed)". The cheat is in the Gameplay submenu.

### CODE REVIEW

**Comment by epic-piano** [comment]
It would also be awesome if in single player you could just automatically equip a second weapon (like a pistol or CMP150) by running over a second one. I don't know why the game only allows for dual weapons when you kill a character who dual wielded them.

**Comment by throwaway29292** [comment]
It might've done so w/o the cheat (not sure right now), but now i can just run over two dead people with the gun and i'll get dualwield.

**Comment by epic-piano** [comment]
> It might've done so w/o the cheat (not sure right now), but now i can just run over two dead people with the gun and i'll get dualwield.

Are you absolutely sure??? First level - I killed the first 10 guards who all equip CMP150's and yet it would still only let me pull out a single CMP150. It's only if you kill a guard who himself fires akimbo that upon picking up both his weapons allows you to dual-wield... so for instance - I can pick up the dual Silenced Pistols from the guard next to the programmer because he wields both the weapons, but I can't pick up dual CMP150's despite almost every single guard in that level holding one. 

It couldn't be something simple like once a second weapons is picked up it enables dual wield for that weapon?

Also, some of the animations seem a bit buggered in Dual-weilding the Sniper rifles... they're very flickery and some of the weapons show Jo with 4 hands when reloading (the Dragon, K7 Avenger, Shotgun, Crossbow etc)... just FYI.

**Comment by fgsfdsfgs** [comment]
Yeah, some guns were not meant to be dual wielded, obviously. But I guess that's part of the fun. With the cheat on I can dual wield CMPs in Defection just fine by picking one up from the first guard and then another one from the second guard, so it looks to me like it works.

**Comment by epic-piano** [comment]
Oh... I meant getting the Dual wield to work without the cheat - I mean, to me it makes sense that if you pick up one gun, then immediately get another same gun you should be able to dual wield automatically. Even if a person doesn't want to dual wield, they can easily put the 2nd gun away and just one.

Just makes it easier than having to use the cheat - not that it's a big deal but figured it was like a quality of life adjustment to the game.

Also, I noticed that there is now an option for Detail textures... is that for a possible Detail Alpha texture overlay on the original textures, or are there some nice juicy High-resolution textures coming in the near future?
________________________________

**Comment by fgsfdsfgs** [comment]
Ah, yeah. That might be a good idea for an extended gameplay option at some point in the future.

The detail textures option is for the original N64 detail textures, it's not related to HD textures. It generally adds some noise to some textures when you get up close. Most visible on some Grid walls and the Defection carpets.

**Comment by throwaway29292** [comment]
Having hires textures would be cool. I already tried upscaling some
oroginal pd textures with stable diffusion AI and got some quite nice
results.

That's offtopic tho. Anyway, thanks for marging my pr.

Dne út 14. 5. 2024 11:48 uživatel fgsfds ***@***.***> napsal:

**Comment by Atari-Dude** [comment]
There are a _lot_ of visual bugs/oddities with this cheat currently, though, which should be ironed out at some point...

Hands are obviously in illogical places, but look especially funky when doing the original full reloading animations lol. Some weapons are off-centered, like the left K7 Avenger is placed nearly in the center of the screen. Some weapons look really odd with the models not mirrored, like the Laptop Gun. Some just don't render at all, like the Devastator, so you just fire out of an invisible weapon.

All newly dual-wieldable weapons should do a generic reload off-screen like the Phoenix does while dual-wielded.

File: src/game/inv.c [hunk 1]
```diff
@@ -510,6 +510,17 @@ s32 invGiveWeaponsByProp(struct prop *prop)
 			weapon = prop->weapon;
 			weaponnum = weapon->weaponnum;
 			otherweaponnum;
+#ifndef PLATFORM_N64
+			//This piece of code makes the cheat work in singleplayer. 
+			if (cheatIsActive(CHEAT_DUALWIELDALLGUNS) && !g_Vars.normmplayerisrunning)
```
Comment (blaiz) [inline src/game/inv.c:515]: If I understand this line correctly, this prevent the cheat from working in combat simulator, is that right? What's the reason for that? I'm asking cause I'm working on making the "All Guns" cheat work for multiple players, not just solo, and I'm wondering whether it should work just in Co-Op/Counter-Op or also in Combat Simulator.

**Pull request merged**